compiler/rustc_target/src/target_features.rs

```diff
@@ -482,6 +482,39 @@ const MIPS_FEATURES: &[(&str, Stability, ImpliedFeatures)] = &[
     // tidy-alphabetical-end
 ];
 
+const NVPTX_FEATURES: &[(&str, Stability, ImpliedFeatures)] = &[
+    // tidy-alphabetical-start
+    ("sm_20", Unstable(sym::nvptx_target_feature), &[]),
+    ("sm_21", Unstable(sym::nvptx_target_feature), &["sm_20"]),
+    ("sm_30", Unstable(sym::nvptx_target_feature), &["sm_21"]),
+    ("sm_32", Unstable(sym::nvptx_target_feature), &["sm_30"]),
+    ("sm_35", Unstable(sym::nvptx_target_feature), &["sm_32"]),
+    ("sm_37", Unstable(sym::nvptx_target_feature), &["sm_35"]),
+    ("sm_50", Unstable(sym::nvptx_target_feature), &["sm_37"]),
+    ("sm_52", Unstable(sym::nvptx_target_feature), &["sm_50"]),
+    ("sm_53", Unstable(sym::nvptx_target_feature), &["sm_52"]),
+    ("sm_60", Unstable(sym::nvptx_target_feature), &["sm_53"]),
+    ("sm_61", Unstable(sym::nvptx_target_feature), &["sm_60"]),
+    ("sm_62", Unstable(sym::nvptx_target_feature), &["sm_61"]),
+    ("sm_70", Unstable(sym::nvptx_target_feature), &["sm_62"]),
+    ("sm_72", Unstable(sym::nvptx_target_feature), &["sm_70"]),
+    ("sm_75", Unstable(sym::nvptx_target_feature), &["sm_72"]),
+    ("sm_80", Unstable(sym::nvptx_target_feature), &["sm_75"]),
+    ("sm_86", Unstable(sym::nvptx_target_feature), &["sm_80"]),
+    ("sm_87", Unstable(sym::nvptx_target_feature), &["sm_86"]),
+    ("sm_90", Unstable(sym::nvptx_target_feature), &["sm_87"]),
+    ("sm_90a", Unstable(sym::nvptx_target_feature), &["sm_90"]),
+    // tidy-alphabetical-end
+    // tidy-alphabetical-start
+    ("sm_100", Unstable(sym::nvptx_target_feature), &["sm_90a"]),
+    ("sm_100a", Unstable(sym::nvptx_target_feature), &["sm_100"]),
+    ("sm_101", Unstable(sym::nvptx_target_feature), &["sm_100a"]),
+    ("sm_101a", Unstable(sym::nvptx_target_feature), &["sm_101"]),
+    ("sm_120", Unstable(sym::nvptx_target_feature), &["sm_101a"]),
+    ("sm_120a", Unstable(sym::nvptx_target_feature), &["sm_120"]),
```

Comment:
I think it is better to comment these out until rust supports the required LLVM version.

```suggestion
    // TODO: requires LLVM 21+
    // ("sm_100", Unstable(sym::nvptx_target_feature), &["sm_90a"]),
    // ("sm_100a", Unstable(sym::nvptx_target_feature), &["sm_100"]),
    // ("sm_101", Unstable(sym::nvptx_target_feature), &["sm_100a"]),
    // ("sm_101a", Unstable(sym::nvptx_target_feature), &["sm_101"]),
    // ("sm_120", Unstable(sym::nvptx_target_feature), &["sm_101a"]),
    // ("sm_120a", Unstable(sym::nvptx_target_feature), &["sm_120"]),
```

Reply:
These are in LLVM-20, which is what Rust's `src/llvm-project` submodule uses (and thus I think all the official builds). The failing CI build environment is using LLVM-18, but I don't know what the convention is for such backward compatibility. I'm fine just commenting these out for now.

@gonzalobg Is it correct for these features to represent a total order, or is there a more general DAG?

Reply:
The non `_a` features extend each other: `sm_90` is a subset of `sm_100`.

The `_a` features contain some unique functionality: `sm_90a` is not a sub-set of `sm_100a`.

Reply:
I added all the `ptx*` features and some documentation. I've tested that one can use `-Ctarget-cpu=sm_62 -Ctarget-feature=+ptx60` and then successfully run a kernel that uses `barrier.sync` (which requires `sm_70` and `ptx60`) on later architectures. The presence of `#[target_feature(enable = "sm_70", enable = "ptx60")]` is not sufficient to increase the PTX version, so it'll be a linker error if you elide the `-Ctarget-feature=+ptx60` above. As such, one will probably want to guard such intrinsics with `#[cfg(target_feature = "ptx60")]`.

Reply:
I'll follow up with a PR to stdarch and will revise docs there. We don't currently have nvptx intrinsics that are gated on target features so I'm not sure how to add the tests you suggest in this repo without replicating some of what is coming for stdarch.